linera-chain/src/block_tracker.rs

```diff
@@ -261,6 +229,44 @@ impl<'resources, 'blobs> BlockExecutionTracker<'resources, 'blobs> {
         Ok(())
     }
 
+    async fn execute_operation<C>(
+        &mut self,
+        chain: &mut ExecutionStateView<C>,
+        operation: &Operation,
+        txn_tracker: &mut TransactionTracker,
+    ) -> Result<(), ChainError>
+    where
+        C: Context + Clone + Send + Sync + 'static,
+        C::Extra: ExecutionRuntimeContext,
+    {
+        let execution_context = ChainExecutionContext::Operation(txn_tracker.transaction_index());
+        self.resource_controller_mut()
+            .with_state(&mut chain.system)
+            .await?
+            .track_operation(operation)
+            .with_execution_context(execution_context)?;
```

Comment:
This piece was previously done after we execute the operation but doing it early could detect cases where the operation couldn't have been paid for.